File: app/src/main/kotlin/tech/mmarca/openvitals/core/period/PeriodCalculations.kt
package tech.mmarca.openvitals.core.period

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.temporal.TemporalAdjusters

fun periodFor(
    range: TimeRange,
    anchorDate: LocalDate,
    today: LocalDate = LocalDate.now(),
    weekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
    clipCurrentWeekToToday: Boolean = true,
): DatePeriod = when (range) {
    TimeRange.DAY -> DatePeriod(
        start = anchorDate,
        end = anchorDate,
    )

    TimeRange.WEEK -> weekPeriodFor(
        anchorDate = anchorDate,
        today = today,
        weekPeriodMode = weekPeriodMode,
        clipCurrentWeekToToday = clipCurrentWeekToToday,
    )

    TimeRange.MONTH -> {
        val start = anchorDate.withDayOfMonth(1)
        val end = anchorDate.withDayOfMonth(anchorDate.lengthOfMonth()).coerceAtMost(today)
        DatePeriod(start = start, end = end)
    }

    TimeRange.YEAR -> {
        val start = anchorDate.withDayOfYear(1)
        val end = anchorDate.withDayOfYear(anchorDate.lengthOfYear()).coerceAtMost(today)
        DatePeriod(start = start, end = end)
    }
}

fun displayPeriodFor(
    range: TimeRange,
    anchorDate: LocalDate,
    today: LocalDate = LocalDate.now(),
    weekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
): DatePeriod =
    periodFor(
        range = range,
        anchorDate = anchorDate,
        today = today,
        weekPeriodMode = weekPeriodMode,
        clipCurrentWeekToToday = false,
    )

fun previousPeriodFor(
    range: TimeRange,
    anchorDate: LocalDate,
    today: LocalDate = LocalDate.now(),
    weekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
): DatePeriod =
    PeriodSelection(
        selectedRange = range,
        selectedDate = anchorDate.coerceAtMost(today),
    ).previousPeriod().period(today, weekPeriodMode)

fun baselinePeriodBefore(
    period: DatePeriod,
    days: Long = 90,
): DatePeriod =
    DatePeriod(
        start = period.start.minusDays(days),
        end = period.start.minusDays(1),
    )

fun periodWindowsFor(
    range: TimeRange,
    anchorDate: LocalDate,
    today: LocalDate = LocalDate.now(),
    baselineDays: Long = DefaultBaselineDays,
    weekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
): PeriodWindows {
    val current = periodFor(range, anchorDate.coerceAtMost(today), today, weekPeriodMode)
    val previous = previousPeriodFor(range, anchorDate, today, weekPeriodMode)
    val baseline = baselinePeriodBefore(current, baselineDays)
    return PeriodWindows(
        current = current,
        previous = previous,
        baseline = baseline,
    )
}

internal fun TimeRange.shift(anchorDate: LocalDate, steps: Long): LocalDate = when (this) {
    TimeRange.DAY -> anchorDate.plusDays(steps)
    TimeRange.WEEK -> anchorDate.plusWeeks(steps)
    TimeRange.MONTH -> anchorDate.plusMonths(steps)
    TimeRange.YEAR -> anchorDate.plusYears(steps)
}

private fun weekPeriodFor(
    anchorDate: LocalDate,
    today: LocalDate,
    weekPeriodMode: WeekPeriodMode,
    clipCurrentWeekToToday: Boolean,
): DatePeriod =
    when (weekPeriodMode) {
        WeekPeriodMode.MONDAY_TO_SUNDAY -> {
            val start = anchorDate.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
            DatePeriod(
                start = start,
                end = if (clipCurrentWeekToToday) {
                    start.plusDays(6).coerceAtMost(today)
                } else {
                    start.plusDays(6)
                },
            )
        }
        WeekPeriodMode.LAST_7_DAYS -> DatePeriod(
            start = anchorDate.minusDays(6),
            end = anchorDate,
        )
    }

internal const val DefaultBaselineDays = 90L

File: lib/core/period/period_calculations.dart
import '../time/local_date.dart';
import 'period_load_query.dart';
import 'period_selection.dart';
import 'time_range.dart';

const int defaultBaselineDays = 90;

extension TimeRangeShift on TimeRange {
  LocalDate shift(LocalDate anchorDate, int steps) {
    switch (this) {
      case TimeRange.day:
        return anchorDate.plusDays(steps);
      case TimeRange.week:
        return anchorDate.plusWeeks(steps);
      case TimeRange.month:
        return anchorDate.plusMonths(steps);
      case TimeRange.year:
        return anchorDate.plusYears(steps);
    }
  }
}

DatePeriod periodFor(
  TimeRange range,
  LocalDate anchorDate, {
  LocalDate? today,
  WeekPeriodMode weekPeriodMode = WeekPeriodMode.mondayToSunday,
  bool clipCurrentWeekToToday = true,
}) {
  final resolvedToday = today ?? LocalDate.now();
  switch (range) {
    case TimeRange.day:
      return DatePeriod(anchorDate, anchorDate);
    case TimeRange.week:
      return _weekPeriodFor(
        anchorDate,
        resolvedToday,
        weekPeriodMode,
        clipCurrentWeekToToday,
      );
    case TimeRange.month:
      final start = anchorDate.withDayOfMonth(1);
      final end = anchorDate
          .withDayOfMonth(anchorDate.lengthOfMonth)
          .coerceAtMost(resolvedToday);
      return DatePeriod(start, end);
    case TimeRange.year:
      final start = anchorDate.withDayOfYear(1);
      final end = anchorDate
          .withDayOfYear(anchorDate.lengthOfYear)
          .coerceAtMost(resolvedToday);
      return DatePeriod(start, end);
  }
}

DatePeriod displayPeriodFor(
  TimeRange range,
  LocalDate anchorDate, {
  LocalDate? today,
  WeekPeriodMode weekPeriodMode = WeekPeriodMode.mondayToSunday,
}) =>
    periodFor(
      range,
      anchorDate,
      today: today,
      weekPeriodMode: weekPeriodMode,
      clipCurrentWeekToToday: false,
    );

DatePeriod previousPeriodFor(
  TimeRange range,
  LocalDate anchorDate, {
  LocalDate? today,
  WeekPeriodMode weekPeriodMode = WeekPeriodMode.mondayToSunday,
}) {
  final resolvedToday = today ?? LocalDate.now();
  return PeriodSelection(range, anchorDate.coerceAtMost(resolvedToday))
      .previousPeriod()
      .period(today: resolvedToday, weekPeriodMode: weekPeriodMode);
}

DatePeriod baselinePeriodBefore(DatePeriod period, {int days = 90}) =>
    DatePeriod(period.start.minusDays(days), period.start.minusDays(1));

PeriodWindows periodWindowsFor(
  TimeRange range,
  LocalDate anchorDate, {
  LocalDate? today,
  int baselineDays = defaultBaselineDays,
  WeekPeriodMode weekPeriodMode = WeekPeriodMode.mondayToSunday,
}) {
  final resolvedToday = today ?? LocalDate.now();
  final current = periodFor(
    range,
    anchorDate.coerceAtMost(resolvedToday),
    today: resolvedToday,
    weekPeriodMode: weekPeriodMode,
  );
  final previous = previousPeriodFor(
    range,
    anchorDate,
    today: resolvedToday,
    weekPeriodMode: weekPeriodMode,
  );
  final baseline = baselinePeriodBefore(current, days: baselineDays);
  return PeriodWindows(
    current: current,
    previous: previous,
    baseline: baseline,
  );
}

DatePeriod _weekPeriodFor(
  LocalDate anchorDate,
  LocalDate today,
  WeekPeriodMode weekPeriodMode,
  bool clipCurrentWeekToToday,
) {
  switch (weekPeriodMode) {
    case WeekPeriodMode.mondayToSunday:
      final start = anchorDate.previousOrSame(DateTime.monday);
      final rawEnd = start.plusDays(6);
      final end = clipCurrentWeekToToday ? rawEnd.coerceAtMost(today) : rawEnd;
      return DatePeriod(start, end);
    case WeekPeriodMode.last7Days:
      return DatePeriod(anchorDate.minusDays(6), anchorDate);
  }
}
